algorithms/kernel/svm/oneapi/svm_train_cache_oneapi.h

```diff
@@ -0,0 +1,239 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+/*
+//++
+//  SVM cache structure implementation
+//--
+*/
+
+#ifndef __SVM_TRAIN_CACHE_ONEAPI_H__
+#define __SVM_TRAIN_CACHE_ONEAPI_H__
+
+#include "service/kernel/service_utils.h"
+#include "externals/service_memory.h"
+#include "service/kernel/data_management/service_micro_table.h"
+#include "service/kernel/data_management/service_numeric_table.h"
+#include "data_management/data/numeric_table_sycl_homogen.h"
+#include "algorithms/kernel/svm/oneapi/svm_helper_oneapi.h"
+
+using namespace daal::services::internal;
+using namespace daal::data_management;
+
+namespace daal
+{
+namespace algorithms
+{
+namespace svm
+{
+namespace training
+{
+namespace internal
+{
+/**
+ * Types of caches for kernel function values
+ */
+enum SVMCacheOneAPIType
+{
+    noCache,     /*!< No storage for caching kernel function values is provided */
+    simpleCache, /*!< Storage for caching ALL kernel function values is provided */
+    lruCache     /*!< Storage for caching PART of kernel function values is provided;
+                         LRU algorithm is used to exclude values from cache */
+};
+
+/**
+ * Common interface for cache that stores kernel function values
+ */
+template <typename algorithmFPType>
+class SVMCacheOneAPIIface
+{
+public:
+    virtual ~SVMCacheOneAPIIface() {}
+
+    virtual services::Status compute(const NumericTablePtr & xTable, const services::Buffer<int> & wsIndices, const size_t p) = 0;
+
+    virtual const services::Buffer<algorithmFPType> & getSetRowsBlock() const = 0;
+    virtual services::Status copyLastToFirst()                                = 0;
+
+protected:
+    SVMCacheOneAPIIface(const size_t blockSize, const size_t lineSize, const kernel_function::KernelIfacePtr & kernel)
+        : _lineSize(lineSize), _blockSize(blockSize), _kernel(kernel)
+    {}
+
+    const size_t _lineSize;                        /*!< Number of elements in the cache line */
+    const size_t _blockSize;                       /*!< Number of cache lines */
+    const kernel_function::KernelIfacePtr _kernel; /*!< Kernel function */
+};
+
+template <SVMCacheOneAPIType cacheType, typename algorithmFPType>
+class SVMCacheOneAPI
+{};
+
+/**
+ * No cache: kernel function values are not cached
+ */
+template <typename algorithmFPType>
+class SVMCacheOneAPI<noCache, algorithmFPType> : public SVMCacheOneAPIIface<algorithmFPType>
```

Comment:
Suppose this is not a cache container, but an implementation of kernel functions.

Reply:
The task of the cache is to return the stored kernel functions compute result.